Algo.h

```diff
@@ -129,32 +159,93 @@ class MMP
+        //ax^2 + bx + c = 0
+        double a = alpha*alpha - beta*beta;
+        //γα + 2s1xβ2
+        double b = gamma*alpha + 2.0*ps2.x()*beta*beta;
+        //c = 1/4γ2 − s12β2
+        double c = 0.25*gamma*gamma - ps2.squaredLength()*beta*beta;
+
+        if(b*b - 4*a*c > 0)
+        {
+            double sq_det = sqrt(b*b - 4*a*c); 
+            x = (-b + sq_det)/(2*a);
+        }
+        else
+            x = -1.0;
 
         if (x >= st and x <= end) {
             // [st, x] closer to ps1 or ps2?
             if ((st_v - ps1).squaredLength() < (st_v - ps2).squaredLength()) {
-                b_intervals.push_back(Interval(ps1, st, x, i1.ps_d, i1.from, i1.edge));
-                b_intervals.push_back(Interval(ps2, x, end, i2.ps_d, i2.from, i2.edge));
+                b_intervals.push_back(Interval(ps1, st, x, i1.ps_d, i1.from, i1.edge, false));
+                b_intervals.push_back(Interval(ps2, x, end, i2.ps_d, i2.from, i2.edge, false));
             } else {
-                b_intervals.push_back(Interval(ps2, st, x, i2.ps_d, i2.from, i2.edge));
-                b_intervals.push_back(Interval(ps1, x, end, i1.ps_d, i1.from, i1.edge));
+                b_intervals.push_back(Interval(ps2, st, x, i2.ps_d, i2.from, i2.edge, false));
+                b_intervals.push_back(Interval(ps1, x, end, i1.ps_d, i1.from, i1.edge, false));
             }
         } else {
             // the complete interval is near to one single source
             if ((ps2 - st_v).squaredLength() < (ps1 - st_v).squaredLength())
-                b_intervals.push_back(Interval(ps2, st, end, i2.ps_d, i2.from, i2.edge));
+                b_intervals.push_back(Interval(ps2, st, end, i2.ps_d, i2.from, i2.edge, false));
             else
-                b_intervals.push_back(Interval(ps1, st, end, i1.ps_d, i1.from, i1.edge));
+                b_intervals.push_back(Interval(ps1, st, end, i1.ps_d, i1.from, i1.edge, false));
         }
 
         return b_intervals;
     }
 
+    int get_next_sane_interval(vector<Interval> &intervals, vector<Interval>::iterator it)
+    {
+        it++;
+        while(it != intervals.end())
+        {
+            auto w = *it;
+            //check if w is too small
+            //w - intervals.begin() gives the index of the current interval
+            if(w.end - w.st > epsilon)
+                return w - intervals.begin();
+            it++;
+        }
+        return -1;
+    }
+
     vector<Interval> sanitize_and_merge(vector<Interval> &intervals)
     {
         // FILL
         // remove too small intervals, merge intervals with same pos and ps_d etc.
-        return {};
+        vector<Interval> sanitized_intervals;
+        for (auto it = intervals.begin(); it != intervals.end(); it++)
+        {
+            auto w = *it;
+            //check if w is too small
+            if(w.end - w.st > epsilon)
+                sanitized_intervals.push_back(w);
```

Comment:
You have to merge any two adjacent intervals with same ps_d & pos, and remove any small intervals. I do not think you require a helper function for that. Just keep two iterators.